srml/support/src/traits.rs

```diff
@@ -640,3 +640,30 @@ pub trait ChangeMembers<AccountId> {
 impl<T> ChangeMembers<T> for () {
 	fn change_members(_incoming: &[T], _outgoing: &[T], _new_set: &[T]) {}
 }
+
+/// A generic trait for reporting slashing violations.
+pub trait ReportSlash<Misbehaved, Reporters, Hash> {
+	/// Reports slashing for a misconduct where `to_slash` are the misbehaved entities and
+	/// `to_reward` are the entities that detected and reported the misbehavior
+	///
+	/// Returns `Ok` if the misconduct was unique otherwise `Err`
+	fn slash(to_slash: Misbehaved, to_reward: Reporters, footprint: Hash) -> Result<(), ()>;
+}
+
+/// A generic trait for enacting slashes.
+pub trait DoSlash<Misbehaved, Reporters, Severity, Kind> {
+	/// Performs the actual slashing and rewarding based on severity
+	fn do_slash(to_slash: Misbehaved, to_reward: Reporters, severity: Severity, kind: Kind) -> Result<(), ()>;
```

Comment:
yes, this is not an elegant abstraction but because we need to keep track on severity for a specific kind, I don't see another option when we need to take the history into account.